docs/exts/taglist.py

```diff
@@ -0,0 +1,112 @@
+from sphinx.util import logging
+from docutils.parsers.rst import Directive
+from docutils import nodes
+from sphinx.environment import NoUri
+
+def initialize(app):
+    app.builder.env.tags = dict()
+
+class tag_node(nodes.General, nodes.Element): pass
+class taglist(nodes.General, nodes.Element):
+    tags = []
+
+class TagDirective(Directive):
+    '''
+    Directive for adding tags to a document
+    '''
+    has_content = False
+    required_arguments = 0
+    optional_arguments = 25
+
+    def run(self):
+        env = self.state.document.settings.env
+
+        tag = tag_node()
+        tag.tags = [tagname.strip().replace(',', '') for tagname in self.arguments]
+
+        return [tag]
+
+class TagListDirective(Directive):
+    '''
+    Directive for generating a list of tagged articles
+    '''
+    required_arguments = 0
+    optional_arguments = 25
+
+    def run(self):
+        tl = taglist('')
+        tl.tags = self.arguments
+
+        return [tl]
+
+def process_tags(app, doctree):
+    '''
+    Traverse all tag nodes and save metadata. Finally remove node
+    from the doctree as it's not supposed to render.
+    '''
+    env = app.builder.env
+
+    for node in doctree.traverse(tag_node):
+        for tagname in node.tags:
+            if tagname not in env.tags:
+                env.tags[tagname] = []
+            env.tags[tagname].append({
+                'docname': env.docname,
+                'title': env.titles[env.docname][0].rawsource
+            })
+        node.replace_self([])
+
+def process_taglist_nodes(app, doctree, fromdocname):
```

Comment:
I think this should be broken up. Perhaps even abstract away the business logic into a class which this just uses, for future extendability and maintenance reasons.